mesonbuild/compilers/compilers.py

```diff
@@ -915,7 +905,15 @@ def get_gcc_soname_args(gcc_type, prefix, shlib_name, suffix, path, soversion, i
         if soversion is not None:
             install_name += '.' + soversion
         install_name += '.dylib'
-        return ['-install_name', os.path.join('@rpath', install_name)]
+        args = ['-install_name', os.path.join(libdir, install_name)]
+        if version and len(version.split('.')) == 3:
+            splitted = version.split('.')
+            major = int(splitted[0])
+            minor = int(splitted[1])
+            revision = int(splitted[2])
+            args += ['-compatibility_version', '%d' % (major + minor + 1)]
+            args += ['-current_version', '%d.%d' % (major + minor + 1, revision)]
```

Comment:
I have not looked at the meson code before, so I apologize if I am misunderstanding something that's obvious to everyone else, but why `major + minor + 1`? This would mean that if I have a library where `version` is, say, `2.5.9`, then the `-compatibility_version` would be 2 + 5 + 1 = `8` (and `-current_version` would be `8.9`). Then if a later version of the library is released where `version` is `3.0.0`, `-compatibility_version` would be 3 + 0 + 1 = `4` and `-current_version` would be `4.0`. Is that what you want?

Note that current version and compatibility version are always a three-part version number. You don't have to specify all of the parts of course; they'll be zero if you don't specify them.

The current version and compatibility version are secondary to the library's major version number, which is in the filename and in the `install_name`. So it's perfectly reasonable for a project to go from libfoo.1.dylib with current version 42.3.0 to libfoo.2.dylib with current version 1.0.0. (On macOS, the library major version doesn't even have to be a number; you could just as easily upgrade from libfoo.A.dylib to libfoo.B.dylib.) Since the `install_name` has changed, everything using the library would need to be recompiled anyway, and the version comparisons against the previous library are irrelevant.

I don't know whether in meson the library major version is part of `version` or separate from it. If the library major version is part of `version` (i.e. if `version` being `3.0.0` means that the filename is `libfoo.3.dylib`), then what you're doing would work, but would be strange. (Why start the versioning for libfoo.3.dylib at 4.0.0? Why not 1.0.0?) If the library major version is separate from `version` (e.g. if you can separately specify that `version` is `3.0.0` and the library major version is 5), then what you're doing is wrong. If my hypothetical project's library major version didn't change from version 2.5.9 to 3.0.0, then a program linked with my 2.5.9 library believes it requires compatibility version 8 and later, but when my library is upgraded to 3.0.0, then my library will only claim to provide compatibility version 4 and the program will not run, saying the library is too old.

What is `version`? Is it the version of the entire project, or the version of an individual library? If it's the version of the project, that may be wrong, since a project may contain more than one library, and there need not be any relationship between the version numbering scheme of the project and the version numbering scheme of its libraries. If it's the version of a library, then why munge it at all? Just set `-current_version` to `version` and set `-compatibility_version` to the first digit of `version`.

One caution though: the compatibility version should never be zero. Very old versions of Apple's linker don't allow that, and the link will fail. The `+ 1` as written in the PR would prevent that problem, but directly using the first digit of a project-specified `version` might not work, if the developer didn't know about this problem and started their versions at 0.x. If `version` is a library version, then maybe the best thing to do is to use the value, but print a warning if it starts with 0.

Computing the compatibility version might also be wrong. My Mac's system library, /usr/lib/libSystem.B.dylib, has a current version of 1238.60.2, but its compatibility version is 1.0.0. You may need to allow the developer to specify their libraries' compatibility versions. I don't know if these concepts have analogs on other operating systems...

Reply:
> Then if a later version of the library is released where version is 3.0.0, -compatibility_version would be 3 + 0 + 1 = 4 and -current_version would be 4.0. Is that what you want?

I think so, yes. At least thats exactly the same that seems to happen when I use autotools libtool, not sure about CMake.

> What is version? Is it the version of the entire project, or the version of an individual library?

The individual library version.

Reply:
> I think so, yes. At least thats exactly the same that seems to happen when I use autotools libtool, not sure about CMake.

Well [libtool treats](https://autotools.io/libtool/version.html) its `-version-info` parameter as  `current`:`revision`:`age`, which is not a traditional `major`.`minor`.`revision` version number. libtool computes the `major`.`minor`.`revision` version number from the `current`:`revision`:`age`. Does meson treat its library `version` the same as libtool's `-version-info`? That's not mentioned in the [meson documentation](http://mesonbuild.com/Reference-manual.html#shared_library).